In [1]:
import pandas as pd
from pathlib import Path
from collections import Counter
from konlpy.tag import Okt

In [2]:
path = Path.cwd()
path = str(path)+'/GFMS DB/'
path

'/Users/keumhyeonjun/Projects/행안부/자재/GFMS DB/'

# 테이블 join > ~~~장애에 사용된 .... 장비의 정보 보기

## 작업지시결과

In [3]:
fmsa10=pd.read_excel(path+'FMSA10_작업지시_결과.xlsx')
fmsa10.head()


comp_cd                 job_cd  job_tpy                         job_nm  \
0    SEJO           211001-00060        5             1동 307호 벽에 못 제거 요청   
1    SEJO           211001-00053        5  1동 411호 천장 고양이 구조로 인한 텍스분리 요청   
2    SEJO  M_202110011345461_001        3            5동 정문 도로 중앙규제봉 교체작업   
3    SEJO  M_202110011511311_001        3   SCC홍보동 부출입구쪽 외부창옆 바닥 보도블럭 침하   
4    SEJO           211001-00010        5      2동 438호 옆 오른쪽 1번 칸 양변기 막힘   

   job_stat job_jisi_dt   job_enddt                 job_shm  \
0         4  2021-10-01  2021-10-01   2021-10-01 오후 3:59:00   
1         4  2021-10-01  2021-10-01   2021-10-01 오후 3:30:00   
2         4  2021-10-01  2021-10-01   2021-10-01 오전 9:45:00   
3         4  2021-10-01  2021-10-01   2021-10-01 오후 1:23:00   
4         4  2021-10-01  2021-10-01  2021-10-01 오전 10:38:00   

                  job_ehm job_cause  ... work_time bosu_gyuljae_yn  emc_yn  \
0   2021-10-01 오후 4:34:00       NaN  ...       NaN               Y       N   
1   2021-10-01 오후 3:48:00       NaN  ...       NaN               Y       N   
2  2021-10-01 오전 10:37:00       NaN  ...       NaN               Y       N   
3   2021-10-01 오후 2:34:00       NaN  ...       NaN               Y       N   
4   2021-10-01 오후 2:20:00       NaN  ...       NaN               N       N   

  const_cd  as_class_lv1 as_class_lv2  common_wrk  emc_id copor_yn kukmin_yn  
0      NaN           NaN          NaN         NaN     NaN        N         N  
1      NaN           NaN          NaN         NaN     NaN        N         N  
2      NaN           NaN          NaN         NaN     NaN        N         N  
3      NaN           NaN          NaN         NaN     NaN        N         N  
4      NaN           7.0        701.0         NaN     NaN        N         N  

[5 rows x 61 columns]

## 작업장비이력

In [4]:
fmsa11=pd.read_excel(path+'FMSA11_작업장비이력.xlsx')
len(fmsa11)
fmsa11.head()

comp_cd        job_cd        fac_cd  job_cont             input_sysdt  \
0    SEJO  211001-00002  400201020417       NaN   2021-10-27 오후 3:27:52   
1    SEJO  211001-00003  400201020817       NaN   2021-10-20 오전 9:07:24   
2    SEJO  211001-00008  400101030002       NaN   2021-10-20 오전 9:06:32   
3    SEJO  211001-00012  700201050006       NaN  2021-10-01 오전 10:58:22   
4    SEJO  211001-00013  200401010026       NaN  2021-10-01 오전 10:58:24   

   chg_sysdt input_emp_no  chg_emp_no input_scr_nm  
0        NaN    korea8012         NaN      JOB_A11  
1        NaN    tjsrb3183         NaN      JOB_A11  
2        NaN    tjsrb3183         NaN      JOB_A11  
3        NaN     krearth0         NaN      JOB_A11  
4        NaN     choi3015         NaN      JOB_A11

In [5]:
len(fmsa11)

6831

## 장비마스터

In [6]:
fmsx05=pd.read_excel(path+'FMSX05_장비마스터.xlsx')
print(len(fmsx05))
fmsx05.head()

40811


comp_cd        fac_cd                       fac_nm  class_cd  dept_cd  \
0    SEJO  200201040293  (3)세-1F-0(여)로비-화장실) 세면대 2대   20020104    10201   
1    SEJO  200201010257                   9F여1대변기(5)  20020101    10201   
2    SEJO  200202030001                        우수조#1  20020203    10201   
3    SEJO  200105080599               배기휀#76(EF-504)  20010508    10201   
4    SEJO  200202010001                        시수조#1  20020201    10201   

      fac_model     fac_spec     fac_plc bd_cd fl_cd  ... input_emp_no  \
0  자동(2)/수동(0)          청소수전       (여)로비   NaN   NaN  ...          박남규   
1   CC-113(바닥형)  690*395*435  여-1(동) 화장실   NaN   NaN  ...          최휴영   
2           NaN          NaN         NaN   NaN   NaN  ...          안용준   
3           NaN          NaN         NaN   NaN   NaN  ...          안용준   
4           NaN          NaN         NaN   NaN   NaN  ...          안용준   

  chg_emp_no input_scr_nm use_yn fac_time_yn                cmt del_yn  \
0        NaN    UBIS_F041      Y           N  8002013000158_228      N   
1        NaN    UBIS_F041      Y           N  8002011000172_229      N   
2        NaN    UBIS_F041      Y           N  8002023003021_229      N   
3        안용준    UBIS_F041      Y           N  8001045005090_229      N   
4        NaN    UBIS_F041      Y           N  8002021001021_229      N   

   chk_edit_gb chk_edit_dt  law_fac_yn  
0          NaN         NaN           N  
1          NaN         NaN           N  
2          NaN         NaN           N  
3          NaN         NaN           N  
4          NaN         NaN           N  

[5 rows x 38 columns]

## 작업장비이력 + 장비마스터 join
* 작업에 사용된 장비의 정보 탐색 가능
* 만약 2번 수리된 장비가 있다면?? >> 장비마스터의  input_sysdt 는 최신화??? 아님 장비번호가 2개존재?? >> 작업장비이력의 시간과 장비마스터의 시간 비교해보기

In [7]:
#분석에 필요한 컬럼 가져오기

#작업코드, 장비코드, 인풋시간
fmsa11_1 =fmsa11[['job_cd','fac_cd','input_sysdt']]
#장비코드, 장비명, 인풋시간
fmsx05_1 =fmsx05[['fac_cd','fac_nm','input_sysdt',]] \

#겹치는 컬럼명 변경
fmsx05_1 = fmsx05_1.rename(columns={'input_sysdt':'input_sysdt_03'})
fmsa11_1 = fmsa11_1.rename(columns={'input_sysdt':'input_sysdt_11'})

In [8]:
#데이터프레임 병함
equipment = pd.merge(fmsa11_1, fmsx05_1, how='inner', on='fac_cd')
print(len(equipment))
equipment

6831


job_cd        fac_cd          input_sysdt_11  \
0              211001-00002  400201020417   2021-10-27 오후 3:27:52   
1              211008-00047  400201020417   2021-10-27 오후 3:29:23   
2              211001-00003  400201020817   2021-10-20 오전 9:07:24   
3              211002-00005  400201020817   2021-10-20 오전 9:08:46   
4              220625-00008  400201020817  2022-06-26 오전 12:15:10   
...                     ...           ...                     ...   
6826  M_202206231505243_001  500102050031   2022-06-23 오후 4:44:34   
6827  M_202207011346422_001  200201020170   2022-07-11 오전 9:13:11   
6828  M_202208021424204_001  200201020133   2022-08-02 오후 2:41:33   
6829  M_202208021424204_001  200201020167   2022-08-02 오후 2:41:33   
6830  M_202208221057315_001  200201030273   2022-08-22 오후 4:34:44   

              fac_nm          input_sysdt_03  
0     14동 아날로그감지기 연기  2019-10-30 오전 10:22:38  
1     14동 아날로그감지기 연기  2019-10-30 오전 10:22:38  
2       광전식 스포트형 감지기  2021-03-08 오전 10:59:52  
3       광전식 스포트형 감지기  2021-03-08 오전 10:59:52  
4       광전식 스포트형 감지기  2021-03-08 오전 10:59:52  
...              ...                     ...  
6826           강화도어3   2021-02-18 오후 3:10:35  
6827      3F남1소변기(장)   2019-11-27 오후 4:56:37  
6828      5F남1소변기(1)   2019-11-27 오후 3:37:48  
6829      5F남1소변기(장)   2019-11-27 오후 4:53:41  
6830     3F남1일반비데(3)   2019-11-29 오전 9:17:07  

[6831 rows x 5 columns]

### 여러번 수리되더라도 장비마스터의 입고 시간은 바뀌지 않는다
* 장비가 교체되는 경우 새로운 장비로 등록??
* 장비가 교체되는 경우는 어떻게 되는가??
* 장비는 교체가 아니고 수리만 있는거??

In [9]:
test = pd.DataFrame(equipment['fac_cd'].value_counts())
# 정비이력이 2번이상 존재하는 값 뽑기
test = test[test['fac_cd']>=2]
num = len(test)

In [10]:
cnt = 0
for i in range(num):
    tmp = test['fac_cd'].index[i]
    # 장비마스터의 input시간이 겹치는 경우가 있으면 알리기
    if len(equipment[equipment['fac_cd'] == tmp]['input_sysdt_03'].value_counts()) >= 2 :
        cnt+=1
print('수리가 여러번 된 경우 장비마스터의 장비 시간이 다른경우의 개수 = ',cnt)

수리가 여러번 된 경우 장비마스터의 장비 시간이 다른경우의 개수 =  0


## 작업지시결과,  (작업장비이력 +장비마스터) join
* job_cd의 개수가 더 많은데?? > 이유 찾기

In [11]:
data = pd.merge(fmsa10, equipment, how='left', on='job_cd')
data

comp_cd                 job_cd  job_tpy                         job_nm  \
0        SEJO           211001-00060        5             1동 307호 벽에 못 제거 요청   
1        SEJO           211001-00053        5  1동 411호 천장 고양이 구조로 인한 텍스분리 요청   
2        SEJO  M_202110011345461_001        3            5동 정문 도로 중앙규제봉 교체작업   
3        SEJO  M_202110011511311_001        3   SCC홍보동 부출입구쪽 외부창옆 바닥 보도블럭 침하   
4        SEJO           211001-00010        5      2동 438호 옆 오른쪽 1번 칸 양변기 막힘   
...       ...                    ...      ...                            ...   
20766    SEJO  M_202210110943434_000        3                         출입문 보수   
20767    SEJO  M_202210111009465_000        3          방화문(플로어힌지, 피봇힌지 등) 보수   
20768    SEJO           221011-00057        3     16동 기계실 인버터 판넬 누름 스위치 보수작업   
20769    SEJO           221011-00044        5         16동 1층 동쪽 여자화장실 대변기 막힘   
20770    SEJO           221011-00019        5     17-2동 7층 남자 화장실 대변기 물 계속흐름   

       job_stat job_jisi_dt   job_enddt                 job_shm  \
0             4  2021-10-01  2021-10-01   2021-10-01 오후 3:59:00   
1             4  2021-10-01  2021-10-01   2021-10-01 오후 3:30:00   
2             4  2021-10-01  2021-10-01   2021-10-01 오전 9:45:00   
3             4  2021-10-01  2021-10-01   2021-10-01 오후 1:23:00   
4             4  2021-10-01  2021-10-01  2021-10-01 오전 10:38:00   
...         ...         ...         ...                     ...   
20766         3  2022-10-11  2022-10-11   2022-10-11 오후 1:30:00   
20767         3  2022-10-11  2022-10-11   2022-10-11 오후 1:30:00   
20768         4  2022-10-11  2022-10-11   2022-10-11 오후 3:00:00   
20769         4  2022-10-11  2022-10-11   2022-10-11 오전 8:32:00   
20770         4  2022-10-11  2022-10-11   2022-10-11 오전 8:57:00   

                      job_ehm job_cause  ... as_class_lv1 as_class_lv2  \
0       2021-10-01 오후 4:34:00       NaN  ...          NaN          NaN   
1       2021-10-01 오후 3:48:00       NaN  ...          NaN          NaN   
2      2021-10-01 오전 10:37:00       NaN  ...          NaN          NaN   
3       2021-10-01 오후 2:34:00       NaN  ...          NaN          NaN   
4       2021-10-01 오후 2:20:00       NaN  ...          7.0        701.0   
...                       ...       ...  ...          ...          ...   
20766   2022-10-11 오후 4:00:00       NaN  ...          NaN          NaN   
20767   2022-10-11 오후 4:00:00       NaN  ...          NaN          NaN   
20768   2022-10-11 오후 3:40:00       NaN  ...          NaN          NaN   
20769   2022-10-11 오전 9:30:00       NaN  ...          7.0        701.0   
20770   2022-10-11 오전 9:20:00       NaN  ...          7.0        701.0   

       common_wrk emc_id  copor_yn kukmin_yn        fac_cd  \
0             NaN    NaN         N         N           NaN   
1             NaN    NaN         N         N           NaN   
2             NaN    NaN         N         N           NaN   
3             NaN    NaN         N         N           NaN   
4             NaN    NaN         N         N           NaN   
...           ...    ...       ...       ...           ...   
20766         NaN    NaN         N         N           NaN   
20767         NaN    NaN         N         N           NaN   
20768         NaN    NaN         N         N           NaN   
20769         NaN    NaN         N         N  2.002010e+11   
20770         NaN    NaN         N         N  2.002010e+11   

              input_sysdt_11           fac_nm          input_sysdt_03  
0                        NaN              NaN                     NaN  
1                        NaN              NaN                     NaN  
2                        NaN              NaN                     NaN  
3                        NaN              NaN                     NaN  
4                        NaN              NaN                     NaN  
...                      ...              ...                     ...  
20766                    NaN              NaN                     NaN  
20767                    NaN  

In [12]:
data[data['sjob_cd'].isnull()==False]

comp_cd                 job_cd  job_tpy  \
19       SEJO  M_202110011053510_000        3   
62       SEJO  M_202110010935273_000        3   
63       SEJO  M_202110010937493_000        3   
64       SEJO  M_202110010939252_000        3   
127      SEJO  M_202110050901491_000        3   
...       ...                    ...      ...   
20725    SEJO  M_202210111010164_000        3   
20726    SEJO  M_202210110959223_000        3   
20765    SEJO  M_202210110946542_000        3   
20766    SEJO  M_202210110943434_000        3   
20767    SEJO  M_202210111009465_000        3   

                                     job_nm  job_stat job_jisi_dt   job_enddt  \
19                                 철재난간대 보수         3  2021-10-01  2021-10-01   
62     17-3동 4층 북측계단실 방화문(플로어힌지, 피봇힌지 등) 보수         4  2021-10-01  2021-10-01   
63     17-3동 5층 비상ev실 방화문(플로어힌지, 피봇힌지 등) 보수         4  2021-10-01  2021-10-01   
64     17-3동 6층 비상ev실 방화문(플로어힌지, 피봇힌지 등) 보수         4  2021-10-01  2021-10-01   
127                                   텍스 보수         3  2021-10-05  2021-10-05   
...                                     ...       ...         ...         ...   
20725                                자동문 보수         3  2022-10-11  2022-10-11   
20726                                자동문 보수         3  2022-10-11  2022-10-11   
20765                 방화문(플로어힌지, 피봇힌지 등) 보수         3  2022-10-11  2022-10-11   
20766                                출입문 보수         3  2022-10-11  2022-10-11   
20767                 방화문(플로어힌지, 피봇힌지 등) 보수         3  2022-10-11  2022-10-11   

                     job_shm                job_ehm job_cause  ...  \
19                2021-10-01             2021-10-01       NaN  ...   
62     2021-10-01 오후 1:30:00  2021-10-01 오후 4:00:00       NaN  ...   
63     2021-10-01 오후 1:30:00  2021-10-01 오후 4:00:00       NaN  ...   
64     2021-10-01 오후 1:30:00  2021-10-01 오후 4:00:00       NaN  ...   
127               2021-10-05             2021-10-05       NaN  ...   
...                      ...                    ...       ...  ...   
20725             2022-10-11             2022-10-11       NaN  ...   
20726             2022-10-11             2022-10-11       NaN  ...   
20765  2022-10-11 오후 1:30:00  2022-10-11 오후 4:00:00       NaN  ...   
20766  2022-10-11 오후 1:30:00  2022-10-11 오후 4:00:00       NaN  ...   
20767  2022-10-11 오후 1:30:00  2022-10-11 오후 4:00:00       NaN  ...   

      as_class_lv1 as_class_lv2  common_wrk emc_id  copor_yn kukmin_yn  \
19             NaN          NaN         NaN    NaN         N         N   
62             NaN        401.0         NaN    NaN         N         N   
63             NaN        401.0         NaN    NaN         N         N   
64             NaN        401.0         NaN    NaN         N         N   
127            NaN          NaN         NaN    NaN         N         N   
...            ...          ...         ...    ...       ...       ...   
20725          NaN          NaN         NaN    NaN         N         N   
20726          NaN          NaN         NaN    NaN         N         N   
20765          NaN          NaN         NaN    NaN         N         N   
20766          NaN          NaN         NaN    NaN         N         N   
20767          NaN          NaN         NaN    NaN         N         N   

       fac_cd  input_sysdt_11 fac_nm input_sysdt_03  
19        NaN             NaN    NaN            NaN  
62        NaN             NaN    NaN            NaN  
63        NaN             NaN    NaN            NaN  
64        NaN             NaN    NaN            NaN  
127       NaN             NaN    NaN            NaN  
...       ...             ...    ...            ...  
20725     NaN             NaN    NaN            NaN  
20726     NaN             NaN    NaN            NaN  
20765     NaN             NaN    NaN            NaN  
20766     NaN             NaN    NaN            NaN  
20767     NaN             NaN    NaN            NaN  

[1766 rows x 65 columns]

### join시 fac_cd = null 인 경우 살펴보기
* 작업지시의 장비코드 , 장비데이터의 작업코드 매칭안되는 경우(작업지시의 작업코드에는 있으나 장비데이터의 작업코드는 없음)
* 모든 장비가 등록된 것이 아닌 경우? > 등록된 장비가 수리 될 경우 장비데이터에 기록??

In [13]:
data.columns

Index(['comp_cd', 'job_cd', 'job_tpy', 'job_nm', 'job_stat', 'job_jisi_dt',
       'job_enddt', 'job_shm', 'job_ehm', 'job_cause', 'job_cont', 'job_ilbo',
       'job_kpi', 'job_fac_his', 'job_mng', 'day_night_ind', 'dept_cd',
       'class_cd', 'sjob_cd', 'vend_nm', 'vend_amt', 'as_req_man',
       'as_req_tel', 'as_req_dt', 'as_jubsu', 'input_sysdt', 'chg_sysdt',
       'input_emp_no', 'chg_emp_no', 'input_scr_nm', 'msg_yn', 'as_req_cont',
       'as_req_comp', 'as_happycall_dt', 'as_happycall_man',
       'as_happycall_res1', 'as_happycall_res2', 'as_happycall_score',
       'as_jungyo', 'as_submit', 'as_reg_h', 'as_reg_d', 'gyuljae_no1',
       'gyuljae_no2', 'gyuljae_no3', 'gyuljae_no4', 'gyuljae_no5',
       'gyuljae_yn', 'edit_yn', 'job_doc_yn', 'sjob_jisi_dt', 'work_time',
       'bosu_gyuljae_yn', 'emc_yn', 'const_cd', 'as_class_lv1', 'as_class_lv2',
       'common_wrk', 'emc_id', 'copor_yn', 'kukmin_yn', 'fac_cd',
       'input_sysdt_11', 'fac_nm', 'input_sysdt_03'],
      dt

## join이 되는 경우 살펴보기

In [14]:
equip_work = data[data['fac_cd'].isnull()==False]
equip_work.head()

comp_cd        job_cd  job_tpy                            job_nm  job_stat  \
9     SEJO  211001-00047        5         2동 1구역 430호 남자화장실 비데 전열불량         4   
10    SEJO  211001-00059        5      4동 지하 1층 전기원휴게실내 샤워장 전등점등불량.         4   
12    SEJO  211001-00037        5         4동 608호 정책조정기획관실 전등 점등 불량         4   
15    SEJO  211001-00081        3  1동 1구역 2층 복도내 거실통로유도등 점등불량 교체 작업         4   
16    SEJO  211001-00070        5       2-1동 옥상층 66번 CCTV PTZ 동작 불량         1   

   job_jisi_dt   job_enddt                job_shm                job_ehm  \
9   2021-10-01  2021-10-01  2021-10-01 오후 2:30:00  2021-10-01 오후 2:40:00   
10  2021-10-01  2021-10-01  2021-10-01 오전 9:00:00  2021-10-01 오전 9:30:00   
12  2021-10-01  2021-10-01  2021-10-01 오후 2:25:00  2021-10-01 오후 3:25:00   
15  2021-10-01  2021-10-01  2021-10-01 오후 1:30:00  2021-10-01 오후 1:45:00   
16  2021-10-01  2021-10-01  2021-10-01 오전 9:00:00  2021-10-01 오후 5:31:00   

   job_cause  ... as_class_lv1 as_class_lv2  common_wrk emc_id  copor_yn  \
9        NaN  ...         13.0       1305.0         NaN    NaN         N   
10       NaN  ...         12.0       1201.0         NaN    NaN         N   
12       NaN  ...         12.0       1201.0         NaN    NaN         N   
15       NaN  ...          NaN          NaN         NaN    NaN         N   
16       NaN  ...         19.0       1901.0         NaN    NaN         N   

   kukmin_yn        fac_cd          input_sysdt_11        fac_nm  \
9          N  3.002010e+11   2021-10-01 오후 3:24:08     LN-4E1(E)   
10         N  3.004010e+11  2021-11-24 오전 11:19:20  4동 지하1층 조명기구   
12         N  3.004010e+11   2021-10-01 오후 3:58:15    4동 6층 조명기구   
15         N  4.003010e+11  2021-10-21 오전 10:33:08  1동  거실통로 유도등   
16         N  6.004020e+11   2021-10-01 오후 5:39:22   2동 CCTV #66   

           input_sysdt_03  
9              2019-10-16  
10  2020-05-29 오후 3:40:28  
12  2020-05-29 오후 4:06:09  
15  2019-10-04 오후 4:43:52  
16  2019-10-25 오후 4:30:18  

[5 rows x 65 columns]

## 키워드 추출해서 장비 묶기

In [15]:
equip_work['replaced'] = 0
equip_work['new_fac_nm']=0

/var/folders/nd/cl2nzwh13zx7jdly2t5m95340000gn/T/ipykernel_15955/68209546.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  equip_work['replaced'] = 0
/var/folders/nd/cl2nzwh13zx7jdly2t5m95340000gn/T/ipykernel_15955/68209546.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  equip_work['new_fac_nm']=0


In [16]:
# 새로운 이름 할당 함수
def grouping(df,idx,keyword):
    for i in (keyword):
        if (i in df['fac_nm'][idx]) and (df['replaced'][idx] == 0):
            df['new_fac_nm'][idx] = i
            df['replaced'][idx] = 1

In [17]:
#키워드 추출

# fac_nm_word=''
# for i in range(0,len(equip_work)):
#     tmp =  str(equip_work['fac_nm'].iloc[i])
#     fac_nm_word += '-'+tmp
# okt = Okt()
# nouns = okt.nouns(fac_nm_word) # 명사만 추출
# print(2)
# words = [n for n in nouns if len(n) > 1]
# c=Counter(words)
# keyword_list = c.most_common()

In [18]:
# index초기화
equip_work2 = equip_work.reset_index()
equip_work2.head()

index comp_cd        job_cd  job_tpy                            job_nm  \
0      9    SEJO  211001-00047        5         2동 1구역 430호 남자화장실 비데 전열불량   
1     10    SEJO  211001-00059        5      4동 지하 1층 전기원휴게실내 샤워장 전등점등불량.   
2     12    SEJO  211001-00037        5         4동 608호 정책조정기획관실 전등 점등 불량   
3     15    SEJO  211001-00081        3  1동 1구역 2층 복도내 거실통로유도등 점등불량 교체 작업   
4     16    SEJO  211001-00070        5       2-1동 옥상층 66번 CCTV PTZ 동작 불량   

   job_stat job_jisi_dt   job_enddt                job_shm  \
0         4  2021-10-01  2021-10-01  2021-10-01 오후 2:30:00   
1         4  2021-10-01  2021-10-01  2021-10-01 오전 9:00:00   
2         4  2021-10-01  2021-10-01  2021-10-01 오후 2:25:00   
3         4  2021-10-01  2021-10-01  2021-10-01 오후 1:30:00   
4         1  2021-10-01  2021-10-01  2021-10-01 오전 9:00:00   

                 job_ehm  ... common_wrk emc_id copor_yn  kukmin_yn  \
0  2021-10-01 오후 2:40:00  ...        NaN    NaN        N          N   
1  2021-10-01 오전 9:30:00  ...        NaN    NaN        N          N   
2  2021-10-01 오후 3:25:00  ...        NaN    NaN        N          N   
3  2021-10-01 오후 1:45:00  ...        NaN    NaN        N          N   
4  2021-10-01 오후 5:31:00  ...        NaN    NaN        N          N   

         fac_cd          input_sysdt_11        fac_nm         input_sysdt_03  \
0  3.002010e+11   2021-10-01 오후 3:24:08     LN-4E1(E)             2019-10-16   
1  3.004010e+11  2021-11-24 오전 11:19:20  4동 지하1층 조명기구  2020-05-29 오후 3:40:28   
2  3.004010e+11   2021-10-01 오후 3:58:15    4동 6층 조명기구  2020-05-29 오후 4:06:09   
3  4.003010e+11  2021-10-21 오전 10:33:08  1동  거실통로 유도등  2019-10-04 오후 4:43:52   
4  6.004020e+11   2021-10-01 오후 5:39:22   2동 CCTV #66  2019-10-25 오후 4:30:18   

   replaced new_fac_nm  
0         0          0  
1         0          0  
2         0          0  
3         0          0  
4         0          0  

[5 rows x 68 columns]

#### 키워드 수정은 여기서

In [19]:
import numpy as np

# 키워드 추출 > 키워드 배열에 넣기

# keyword = []
# for i in range(len(keyword_list)):
#     keyword.append(keyword_list[i][0])

# 배열 저장

# keyword =np.array(keyword)
# np.save('keyword_save', keyword) # keyw

In [20]:
#키워드 배열 불러오기
keyword = list(np.load('keyword_save.npy'))

In [21]:
#추출이 잘못된 키워드 수정
keyword.remove('호기')
keyword.remove('화장실')
keyword.remove('시설')
keyword[keyword.index('내기')]='실내기'
keyword[keyword.index('유도')]='유도등'
keyword[keyword.index('스피드')]='스피드게이트'

In [22]:
#새로운 이름 할당
for idx in range(len(equip_work2)):
    grouping(equip_work2,idx,keyword)
    if equip_work2['new_fac_nm'][idx] == 0 :
        equip_work2['new_fac_nm'][idx] = equip_work2['fac_nm'][idx]

/var/folders/nd/cl2nzwh13zx7jdly2t5m95340000gn/T/ipykernel_15955/1006653966.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  equip_work2['new_fac_nm'][idx] = equip_work2['fac_nm'][idx]
/var/folders/nd/cl2nzwh13zx7jdly2t5m95340000gn/T/ipykernel_15955/4217135829.py:6: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['replaced'][idx] = 1


In [23]:
equip_work2

index comp_cd        job_cd  job_tpy  \
0         9    SEJO  211001-00047        5   
1        10    SEJO  211001-00059        5   
2        12    SEJO  211001-00037        5   
3        15    SEJO  211001-00081        3   
4        16    SEJO  211001-00070        5   
...     ...     ...           ...      ...   
6826  20739    SEJO  221011-00011        5   
6827  20759    SEJO  221011-00066        3   
6828  20760    SEJO  221011-00040        3   
6829  20769    SEJO  221011-00044        5   
6830  20770    SEJO  221011-00019        5   

                                      job_nm  job_stat job_jisi_dt  \
0                  2동 1구역 430호 남자화장실 비데 전열불량         4  2021-10-01   
1               4동 지하 1층 전기원휴게실내 샤워장 전등점등불량.         4  2021-10-01   
2                  4동 608호 정책조정기획관실 전등 점등 불량         4  2021-10-01   
3           1동 1구역 2층 복도내 거실통로유도등 점등불량 교체 작업         4  2021-10-01   
4                2-1동 옥상층 66번 CCTV PTZ 동작 불량         1  2021-10-01   
...                                      ...       ...         ...   
6826      13동 5-7 여자화장실 오른쪽 첫번째 대변기 밑 밸브쪽 누수         4  2022-10-11   
6827     13-2동 1층 방재실 R형 수신기 예비전원 불량 교체 (2개)         1  2022-10-11   
6828  꿈샘어린이집 1층 기계실 소화수조A 레벨스위치 1개 작동불량 교체작업         1  2022-10-11   
6829                  16동 1층 동쪽 여자화장실 대변기 막힘         4  2022-10-11   
6830              17-2동 7층 남자 화장실 대변기 물 계속흐름         4  2022-10-11   

       job_enddt                job_shm                 job_ehm  ...  \
0     2021-10-01  2021-10-01 오후 2:30:00   2021-10-01 오후 2:40:00  ...   
1     2021-10-01  2021-10-01 오전 9:00:00   2021-10-01 오전 9:30:00  ...   
2     2021-10-01  2021-10-01 오후 2:25:00   2021-10-01 오후 3:25:00  ...   
3     2021-10-01  2021-10-01 오후 1:30:00   2021-10-01 오후 1:45:00  ...   
4     2021-10-01  2021-10-01 오전 9:00:00   2021-10-01 오후 5:31:00  ...   
...          ...                    ...                     ...  ...   
6826  2022-10-11  2022-10-11 오전 9:40:00  2022-10-11 오전 10:10:00  ...   
6827  2022-10-11  2022-10-11 오후 2:30:00   2022-10-11 오후 3:10:00  ...   
6828  2022-10-11  2022-10-11 오후 1:30:00   2022-10-11 오후 2:30:00  ...   
6829  2022-10-11  2022-10-11 오전 8:32:00   2022-10-11 오전 9:30:00  ...   
6830  2022-10-11  2022-10-11 오전 8:57:00   2022-10-11 오전 9:20:00  ...   

     common_wrk emc_id copor_yn  kukmin_yn        fac_cd  \
0           NaN    NaN        N          N  3.002010e+11   
1           NaN    NaN        N          N  3.004010e+11   
2           NaN    NaN        N          N  3.004010e+11   
3           NaN    NaN        N          N  4.003010e+11   
4           NaN    NaN        N          N  6.004020e+11   
...         ...    ...      ...        ...           ...   
6826        NaN    NaN        N          N  2.002010e+11   
6827        NaN    NaN        N          N  4.002010e+11   
6828        NaN    NaN        N          N  4.004010e+11   
6829        NaN    NaN        N          N  2.002010e+11   
6830        NaN    NaN        N          N  2.002010e+11   

              input_sysdt_11           fac_nm          input_sysdt_03  \
0      2021-10-01 오후 3:24:08        LN-4E1(E)              2019-10-16   
1     2021-11-24 오전 11:19:20     4동 지하1층 조명기구   2020-05-29 오후 3:40:28   
2      2021-10-01 오후 3:58:15       4동 6층 조명기구   2020-05-29 오후 4:06:09   
3     2021-10-21 오전 10:33:08     1동  거실통로 유도등   2019-10-04 오후 4:43:52   
4      2021-10-01 오후 5:39:22      2동 CCTV #66   2019-10-25 오후 4:30:18   
...                      ...              ...                     ...   
6826  2022-10-11 오전 10:38:16        물내림비데 #30   2019-11-26 오후 4:56:29   
6827   2022-10-11 오후 4:09:26    13동 GR형복합식수신기  2019-09-25 오전 10:13:08   
6828   2022-10-11 오후 2:32:34          별동 소화수조   2020-02-27 오후 2:21:25   
6829   2022-10-11 오후 2:59:13  1F여2동쪽 공용대변기(5)   2019-11-27 오후 1:41:05   
6830   2022-10-11 오후 1:42:33       7F남2대변기(3)  2019-11-28 오전 10:45:23   

      replaced   new_fac_nm  
0            0    LN-4E1(E)  
1            1           조명  
2            1           조명  
3            1          유도등  

### 키워드 추출이 되지 않은 항목 > 수작업으로 확인 후 그룹화

In [24]:
# 키워드 개수 확인 함수
def find(df,fac):
    num = 0
    for i in range(len(df)):
        if f'{fac}' in df['index'][i]:
           num += df['fac_nm'][i]
           df = df.drop([i],axis=0)
    print(f'{fac}가 포함된 열은 {num}개 입니다')
    df.reset_index(inplace=True)
    df.drop(['level_0'],axis=1,inplace=True)
    return df


# 바뀌지 않은 장비명 추출
# pd.set_option('display.max_rows',976)
aa = pd.DataFrame(equip_work2[equip_work2['replaced'] == 0]['fac_nm'].value_counts())
aa.reset_index(inplace=True)
aa.colunms = ['equip','num']




/var/folders/nd/cl2nzwh13zx7jdly2t5m95340000gn/T/ipykernel_15955/1464632946.py:18: UserWarning: Pandas doesn't allow columns to be created via a new attribute name - see https://pandas.pydata.org/pandas-docs/stable/indexing.html#attribute-access
  aa.colunms = ['equip','num']


#### 키워드 추출 실패한 장비 수작업 그룹화

In [25]:
a_LT = find(aa,'LT')
a_AHU = find(a_LT,'AHU')
a_NVR = find(a_AHU,'NVR')
a_CCTV = find(a_NVR,'CCTV')
a_급기팬 = find(a_CCTV,'급기팬')

LT가 포함된 열은 231개 입니다
AHU가 포함된 열은 171개 입니다
NVR가 포함된 열은 34개 입니다
CCTV가 포함된 열은 138개 입니다
급기팬가 포함된 열은 25개 입니다


In [26]:
other_keyword = ['LT','AHU','NVR','CCTV','급기팬']
for idx in range(len(equip_work2)):
    grouping(equip_work2,idx,other_keyword)
    if equip_work2['new_fac_nm'][idx] == 0 :
        equip_work2['new_fac_nm'][idx] = equip_work2['fac_nm'][idx]

/var/folders/nd/cl2nzwh13zx7jdly2t5m95340000gn/T/ipykernel_15955/4217135829.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['new_fac_nm'][idx] = i
/var/folders/nd/cl2nzwh13zx7jdly2t5m95340000gn/T/ipykernel_15955/4217135829.py:6: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['replaced'][idx] = 1


# 자재 - 장애 매칭

## FMSA10_작업지시결과 + FMSA11_작업장비이력 + FMSB03_장비마스터
* 총 19000건의 작업자시결과 중 장비 매칭되는 건수 > 5120건

In [27]:
equip_work2

index comp_cd        job_cd  job_tpy  \
0         9    SEJO  211001-00047        5   
1        10    SEJO  211001-00059        5   
2        12    SEJO  211001-00037        5   
3        15    SEJO  211001-00081        3   
4        16    SEJO  211001-00070        5   
...     ...     ...           ...      ...   
6826  20739    SEJO  221011-00011        5   
6827  20759    SEJO  221011-00066        3   
6828  20760    SEJO  221011-00040        3   
6829  20769    SEJO  221011-00044        5   
6830  20770    SEJO  221011-00019        5   

                                      job_nm  job_stat job_jisi_dt  \
0                  2동 1구역 430호 남자화장실 비데 전열불량         4  2021-10-01   
1               4동 지하 1층 전기원휴게실내 샤워장 전등점등불량.         4  2021-10-01   
2                  4동 608호 정책조정기획관실 전등 점등 불량         4  2021-10-01   
3           1동 1구역 2층 복도내 거실통로유도등 점등불량 교체 작업         4  2021-10-01   
4                2-1동 옥상층 66번 CCTV PTZ 동작 불량         1  2021-10-01   
...                                      ...       ...         ...   
6826      13동 5-7 여자화장실 오른쪽 첫번째 대변기 밑 밸브쪽 누수         4  2022-10-11   
6827     13-2동 1층 방재실 R형 수신기 예비전원 불량 교체 (2개)         1  2022-10-11   
6828  꿈샘어린이집 1층 기계실 소화수조A 레벨스위치 1개 작동불량 교체작업         1  2022-10-11   
6829                  16동 1층 동쪽 여자화장실 대변기 막힘         4  2022-10-11   
6830              17-2동 7층 남자 화장실 대변기 물 계속흐름         4  2022-10-11   

       job_enddt                job_shm                 job_ehm  ...  \
0     2021-10-01  2021-10-01 오후 2:30:00   2021-10-01 오후 2:40:00  ...   
1     2021-10-01  2021-10-01 오전 9:00:00   2021-10-01 오전 9:30:00  ...   
2     2021-10-01  2021-10-01 오후 2:25:00   2021-10-01 오후 3:25:00  ...   
3     2021-10-01  2021-10-01 오후 1:30:00   2021-10-01 오후 1:45:00  ...   
4     2021-10-01  2021-10-01 오전 9:00:00   2021-10-01 오후 5:31:00  ...   
...          ...                    ...                     ...  ...   
6826  2022-10-11  2022-10-11 오전 9:40:00  2022-10-11 오전 10:10:00  ...   
6827  2022-10-11  2022-10-11 오후 2:30:00   2022-10-11 오후 3:10:00  ...   
6828  2022-10-11  2022-10-11 오후 1:30:00   2022-10-11 오후 2:30:00  ...   
6829  2022-10-11  2022-10-11 오전 8:32:00   2022-10-11 오전 9:30:00  ...   
6830  2022-10-11  2022-10-11 오전 8:57:00   2022-10-11 오전 9:20:00  ...   

     common_wrk emc_id copor_yn  kukmin_yn        fac_cd  \
0           NaN    NaN        N          N  3.002010e+11   
1           NaN    NaN        N          N  3.004010e+11   
2           NaN    NaN        N          N  3.004010e+11   
3           NaN    NaN        N          N  4.003010e+11   
4           NaN    NaN        N          N  6.004020e+11   
...         ...    ...      ...        ...           ...   
6826        NaN    NaN        N          N  2.002010e+11   
6827        NaN    NaN        N          N  4.002010e+11   
6828        NaN    NaN        N          N  4.004010e+11   
6829        NaN    NaN        N          N  2.002010e+11   
6830        NaN    NaN        N          N  2.002010e+11   

              input_sysdt_11           fac_nm          input_sysdt_03  \
0      2021-10-01 오후 3:24:08        LN-4E1(E)              2019-10-16   
1     2021-11-24 오전 11:19:20     4동 지하1층 조명기구   2020-05-29 오후 3:40:28   
2      2021-10-01 오후 3:58:15       4동 6층 조명기구   2020-05-29 오후 4:06:09   
3     2021-10-21 오전 10:33:08     1동  거실통로 유도등   2019-10-04 오후 4:43:52   
4      2021-10-01 오후 5:39:22      2동 CCTV #66   2019-10-25 오후 4:30:18   
...                      ...              ...                     ...   
6826  2022-10-11 오전 10:38:16        물내림비데 #30   2019-11-26 오후 4:56:29   
6827   2022-10-11 오후 4:09:26    13동 GR형복합식수신기  2019-09-25 오전 10:13:08   
6828   2022-10-11 오후 2:32:34          별동 소화수조   2020-02-27 오후 2:21:25   
6829   2022-10-11 오후 2:59:13  1F여2동쪽 공용대변기(5)   2019-11-27 오후 1:41:05   
6830   2022-10-11 오후 1:42:33       7F남2대변기(3)  2019-11-28 오전 10:45:23   

      replaced new_fac_nm  
0            0  LN-4E1(E)  
1            1         조명  
2            1         조명  
3            1        유도등  
4        

In [28]:
equip_work2['job_cd'].value_counts()

211112-00057    52
220907-00102    48
220825-00057    48
220209-00059    42
211203-00062    42
                ..
220206-00006     1
220204-00027     1
220204-00024     1
220204-00022     1
221011-00019     1
Name: job_cd, Length: 5120, dtype: int64

## FMSA11_작업장비이력 + FMSB07_자재출고
* 약 19000건 중 자재출고와 매칭되는 경우 2336건

In [29]:
fmsb07 = pd.read_excel(path+'FMSB07_자재출고.xlsx')

In [30]:
print(len(fmsb07))
print(len(fmsb07['job_cd'].value_counts()))

5722
4286


In [31]:
jajae_join = pd.merge(fmsa11,fmsb07, on = 'job_cd', how ='inner')
jajae_join

comp_cd_x                 job_cd        fac_cd  job_cont  \
0         SEJO           211001-00003  400201020817       NaN   
1         SEJO           211001-00008  400101030002       NaN   
2         SEJO           211001-00059  300401010012       NaN   
3         SEJO           211001-00062  400101010386       NaN   
4         SEJO           211001-00065  200103010007       NaN   
...        ...                    ...           ...       ...   
3392      SEJO  M_202207011346422_001  200201020170       NaN   
3393      SEJO  M_202207131450110_001  200201020118       NaN   
3394      SEJO  M_202208021424204_001  200201020133       NaN   
3395      SEJO  M_202208021424204_001  200201020167       NaN   
3396      SEJO  M_202208221057315_001  200201030273       NaN   

               input_sysdt_x  chg_sysdt_x input_emp_no_x  chg_emp_no_x  \
0      2021-10-20 오전 9:07:24          NaN      tjsrb3183           NaN   
1      2021-10-20 오전 9:06:32          NaN      tjsrb3183           NaN   
2     2021-11-24 오전 11:19:20          NaN    dlawowls100           NaN   
3      2021-10-20 오전 9:45:37          NaN       dragon72           NaN   
4      2021-10-14 오전 8:35:44          NaN       choi3015           NaN   
...                      ...          ...            ...           ...   
3392   2022-07-11 오전 9:13:11          NaN          as119           NaN   
3393   2022-07-13 오후 3:13:09          NaN          as119           NaN   
3394   2022-08-02 오후 2:41:33          NaN       berdan10           NaN   
3395   2022-08-02 오후 2:41:33          NaN       berdan10           NaN   
3396   2022-08-22 오후 4:34:44          NaN          as119           NaN   

     input_scr_nm_x comp_cd_y  ... gyuljae_no1  gyuljae_no2 gyuljae_no3  \
0           JOB_A11      SEJO  ...         NaN          NaN         NaN   
1           JOB_A11      SEJO  ...         NaN          NaN         NaN   
2           JOB_A11      SEJO  ...         NaN          NaN         NaN   
3           JOB_A11      SEJO  ...         NaN          NaN         NaN   
4           JOB_A11      SEJO  ...         NaN          NaN         NaN   
...             ...       ...  ...         ...          ...         ...   
3392        JOB_A11      SEJO  ...         NaN          NaN         NaN   
3393        JOB_A11      SEJO  ...         NaN          NaN         NaN   
3394      M_JOB_A11      SEJO  ...         NaN          NaN         NaN   
3395      M_JOB_A11      SEJO  ...         NaN          NaN         NaN   
3396        JOB_A11      SEJO  ...         NaN          NaN         NaN   

      gyuljae_no4  gyuljae_no5  gyuljae_yn return_qty mat_check_gb  \
0             NaN          NaN           N        NaN          NaN   
1             NaN          NaN           N        NaN          NaN   
2             NaN          NaN           N        NaN          NaN   
3             NaN          NaN           N        NaN          NaN   
4             NaN          NaN           N        NaN          NaN   
...           ...          ...         ...        ...          ...   
3392          NaN          NaN           N        NaN          NaN   
3393          NaN          NaN           N        NaN          NaN   
3394          NaN          NaN           N        NaN          NaN   
3395          NaN          NaN           N        NaN          NaN   
3396          NaN          NaN           N        NaN          NaN   

      old_mat_cd new_mat_cd  
0            NaN        NaN  
1            NaN        NaN  
2            NaN        NaN  
3            NaN        NaN  
4            NaN        NaN  
...          ...        ...  
3392         NaN        NaN  
3393         NaN        NaN  
3394         NaN        NaN  
3395         NaN        NaN  
3396         NaN        NaN  

[3397 rows x 39 columns]

In [32]:
jajae_join['job_cd'].value_counts()

211103-00012             40
211222-00047             29
211206-00068             16
211230-00047             16
211021-00063             15
                         ..
220303-00022              1
220303-00010              1
220303-00004              1
220302-00046              1
M_202208221057315_001     1
Name: job_cd, Length: 2336, dtype: int64

## FMSA10_작업지시결과 + FMSA11_작업장비이력 + _장비마스터 + FMSB07_자재출고
* 2336 건 >> 작업 + 자재출고 매칭 시 장비 매칭가능

In [33]:
join_data = pd.merge(equip_work2, jajae_join, how = 'inner', on = 'job_cd')

In [34]:
join_data

index comp_cd        job_cd  job_tpy  \
0        10    SEJO  211001-00059        5   
1        15    SEJO  211001-00081        3   
2        16    SEJO  211001-00070        5   
3        16    SEJO  211001-00070        5   
4        58    SEJO  211001-00062        3   
...     ...     ...           ...      ...   
6372  20729    SEJO  221011-00023        3   
6373  20731    SEJO  221011-00037        5   
6374  20759    SEJO  221011-00066        3   
6375  20759    SEJO  221011-00066        3   
6376  20760    SEJO  221011-00040        3   

                                      job_nm  job_stat job_jisi_dt  \
0               4동 지하 1층 전기원휴게실내 샤워장 전등점등불량.         4  2021-10-01   
1           1동 1구역 2층 복도내 거실통로유도등 점등불량 교체 작업         4  2021-10-01   
2                2-1동 옥상층 66번 CCTV PTZ 동작 불량         1  2021-10-01   
3                2-1동 옥상층 66번 CCTV PTZ 동작 불량         1  2021-10-01   
4     9동 701호해양수산사이버안전센터 분말소화기 받침대 노후로 신품교체.         4  2021-10-01   
...                                      ...       ...         ...   
6372     7-2동 7층 남자화장실, 7-1동 5층 남자화장실 대변기 누수         4  2022-10-11   
6373                 11동 1-2 남자화장실 소변기 물 안나옴         4  2022-10-11   
6374     13-2동 1층 방재실 R형 수신기 예비전원 불량 교체 (2개)         1  2022-10-11   
6375     13-2동 1층 방재실 R형 수신기 예비전원 불량 교체 (2개)         1  2022-10-11   
6376  꿈샘어린이집 1층 기계실 소화수조A 레벨스위치 1개 작동불량 교체작업         1  2022-10-11   

       job_enddt                job_shm                 job_ehm  ...  \
0     2021-10-01  2021-10-01 오전 9:00:00   2021-10-01 오전 9:30:00  ...   
1     2021-10-01  2021-10-01 오후 1:30:00   2021-10-01 오후 1:45:00  ...   
2     2021-10-01  2021-10-01 오전 9:00:00   2021-10-01 오후 5:31:00  ...   
3     2021-10-01  2021-10-01 오전 9:00:00   2021-10-01 오후 5:31:00  ...   
4     2021-10-01  2021-10-01 오후 3:30:00   2021-10-01 오후 4:09:00  ...   
...          ...                    ...                     ...  ...   
6372  2022-10-11  2022-10-11 오후 1:30:00   2022-10-11 오후 1:59:00  ...   
6373  2022-10-11  2022-10-11 오전 9:50:00  2022-10-11 오전 10:10:00  ...   
6374  2022-10-11  2022-10-11 오후 2:30:00   2022-10-11 오후 3:10:00  ...   
6375  2022-10-11  2022-10-11 오후 2:30:00   2022-10-11 오후 3:10:00  ...   
6376  2022-10-11  2022-10-11 오후 1:30:00   2022-10-11 오후 2:30:00  ...   

     gyuljae_no1_y gyuljae_no2_y gyuljae_no3_y  gyuljae_no4_y gyuljae_no5_y  \
0              NaN           NaN           NaN            NaN           NaN   
1              NaN           NaN           NaN            NaN           NaN   
2              NaN           NaN           NaN            NaN           NaN   
3              NaN           NaN           NaN            NaN           NaN   
4              NaN           NaN           NaN            NaN           NaN   
...            ...           ...           ...            ...           ...   
6372           NaN           NaN           NaN            NaN           NaN   
6373           NaN           NaN           NaN            NaN           NaN   
6374           NaN           NaN           NaN            NaN           NaN   
6375           NaN           NaN           NaN            NaN           NaN   
6376           NaN           NaN           NaN            NaN           NaN   

      gyuljae_yn_y return_qty  mat_check_gb  old_mat_cd new_mat_cd  
0                N        NaN           NaN         NaN        NaN  
1                Y        NaN           NaN         NaN        NaN  
2                N        NaN           NaN         NaN        NaN  
3                N        NaN           NaN         NaN        NaN  
4                N        NaN           NaN         NaN        NaN  
...            ...        ...           ...         ...        ...  
6372             N        NaN           NaN         NaN        NaN  
6373             N        NaN           NaN         NaN        NaN  
6374             N        NaN           NaN         NaN        NaN  
6375             N        NaN           NaN         NaN        NaN  
6376             N        NaN          

In [35]:
len(join_data['job_cd'].value_counts())

2336

In [36]:
# data = 장비 + 작업지시결과 left join
test = pd.merge(data, fmsb07, how = 'inner', on = 'job_cd')

In [37]:
len(test[test['fac_cd'].isnull() == False]['job_cd'].value_counts())

2336

# EDA

In [38]:
keyword_value = equip_work2['new_fac_nm'].value_counts()

In [39]:
pd.DataFrame(keyword_value)

new_fac_nm
소변기                                488
대변기                                464
펌프                                 360
냉각탑                                349
실내기                                304
...                                ...
1호기 카                                1
14동 F3 ACU                           1
DIGITAL POWER AMP(500W1CH)           1
6동 V1000 #113                        1
OAE-6C3-1(C)                         1

[457 rows x 1 columns]

In [40]:
equip_work2[equip_work2['new_fac_nm']=='승강'][['fac_nm','new_fac_nm','job_nm']]

fac_nm new_fac_nm                          job_nm
14     23호기 승강장         승강      11동 카드리더기로 인한 23호기 파킹해제 불가
74     60호기 승강로         승강         솔비타어린이집 자동구출운전장치 배터리 교체
342    10호기 승강로         승강  10동 음성안내 기판 이상으로 인한 음성안내 기판 교체
344    34호기 승강로         승강       13동 34호기 5층 승장도어 스위치 접점교체
413    25호기 승강로         승강          12동 25호기 카운터 베어링 소음발생 
...         ...        ...                             ...
6259  승강기18호기 카         승강             승강기 내 카내 조명등 1EA 불량
6312    2호기 승강로         승강              9동 2호기 카도어 연동로프 교체
6313   29호기 승강로         승강              13동 29호기 리미트스위치 교체
6495   34호기 승강로         승강               13동 34호기 가이드롤러 조정
6760  승강기13호기 카         승강                   13호기 카 내 등 교체

[172 rows x 3 columns]

In [41]:
pd.options.display.float_format = '{:.0f}'.format

## 장비명은 같은데 코드가 여러개인 경우도 존재
* 그룹화 할 때 유의

In [42]:
equip_work2[equip_work2['fac_nm']=='냉각탑#1'][['job_nm','fac_cd','job_jisi_dt','job_enddt','input_sysdt_03','fac_nm']]

job_nm       fac_cd  \
137                    15-1동 4층 냉각탑 1.2호기 워터가드 투입 (6EA) 200101020022   
3610                      15-1층 4층 냉각탑 냉각수 급수 볼밸브 교체 작업 200101020022   
4014                7-2동 1층 옥외 냉각탑  하부 스트레이너 카바 팩킹 교체작업 200101020020   
4016       11동 옥상 냉각탑1호기 시수공급용 바이패스배관 이음새 누수로 볼밸브,소켓 교체 200101020024   
4335                         11동 옥상 냉동기 냉각탑 플로트밸브 2개 교체 200101020024   
4349                  7동 냉각탑 1,2호기 냉각팬 고열그리스 주입 ( 4개소 ) 200101020020   
4410                                 11동 옥상 냉각탑 수처리제 투입 200101020024   
4416                           15-1동 4층 냉각탑  수처리제 4개 투입 200101020022   
4418                            15-1동 4층 1번 냉각탑  볼탑교체작업 200101020022   
4602                               7동 냉각탑 1,2호기 수처리제 투입 200101020020   
4758                              11동 옥상 냉각탑 수처리제 8통 투입 200101020024   
4877                     7동 옥외 1층 냉각탑1.2호기  냉각수 수처리제 투입 200101020020   
4928                             11동 옥상 냉각탑 수처리제 10통 투입 200101020024   
4932                        15-1동 4층 냉각탑 1.2호기  수처리제 투입 200101020022   
5066                         15-1동 4층 냉각탑 냉각수 처리제 투입 작업 200101020022   
5187                         7-2동 냉각탑 모터 저압진상콘덴서 노후로 교체 200101020020   
5274                                7-2동 냉각탑  수처리제   투입 200101020020   
5282                             11동 옥상 냉각탑 수처리제 10통 투입 200101020024   
5307                           7-2동 1층 냉각탑 1호기 수조 청소 작업 200101020020   
5310                               15동 1,2호 냉각탑 수처리제 투입 200101020022   
5422                             11동 옥상 냉각탑 수처리제 10통 투입 200101020024   
5426                        15-1동 4층 냉각탑 1.2호기 수처리제  투입 200101020022   
5544                         15-1동 4층 냉각탑 1,2호기 수처리제 투입 200101020022   
5709                        15-1동 4층 냉각탑 1,2호 기 수처리제 투입 200101020022   
5805                        7동 냉동기 1,2호기   냉각탑  수처리제 투입 200101020020   
5843                     15-1동 4층 냉각탑 1,2호기 냉각수 수처리제 투입 200101020022   
5982                       7-2동 1층 옥외 냉각탑 1.2호기 수처리제 투입 200101020020   
5989                     15-1동 4층 냉각탑 1,2호기 냉각수 처리제 투입  200101020022   
6009                             11동 옥상 냉각탑 수처리제 10통 투입 200101020024   
6044                         7.8동 옥외 1층 냉각탑 냉각수 수처리제 투입 200101020020   
6101                                  15-1동 4층 냉각탑 수질검사 200101020022   
6126                                    11동 냉각탑 수질검사 의뢰 200101020024   
6128                      15-1동 4층 냉각탑 1,2호기 냉각수 처리제 투입 200101020022   
6246                         15-1동 4층 냉각탑 1,2호기 수처리제 투입 200101020022   
6267                       7-2동 1층 옥외 냉각탑 1.2호기 수처리제 투입 200101020020   
6271                       7-2동 1층 옥외 냉각탑 1.2호기 수처리제 투입 200101020020   
6279  8/22, 8/29 옥상 냉각탑 수처리제 투입 내역 보고(각일별 10개씩 총 20개... 200101020024   
6375                      7-2동 흡수식 냉동기  냉각탑 냉각수 수처리제 투입 200101020020   
6385                             11동 옥상 냉각탑 수처리제 10통 투입 200101020024   
6422                     11동 옥상 냉동기 냉각탑 드레인관 누수로 볼밸브 교체 200101020024   
6430                               15-1동 4층 냉각탑 수처리제 투입 200101020022   
6553                            11동 옥상 냉각탑 수처리제 10통 투입  200101020024   
6562                            15-1동 4층 냉각탑 안 볼탑 교체 작업 200101020022   
6577                        15동 4층 옥상 냉각탑 1.2호기 수처리제 투입 200101020022   
6706                             11동 옥상 냉각탑 수처리제 10통 투입 200101020024   

     job_jisi_dt   job_enddt         input_sysdt_03 fac_nm  
137   2021-10-06  2021-10-06  2019-09-20 오후 1:45:54  냉각탑#1  
3610  2022-04-14  2022-04-14  2019-09-20 오후 1:45:54  냉각탑#1  
4014  2022-05-10  2022-05-10  2019-09-17 오후 3:40:40  냉각탑#1  
4016  2022-05-10  2022-05-10  2019-09-24 오후 4:55:30  냉각탑#1  
4335  2022-05-30  2022-05-30  2019-09-24 오후 4:55:30  냉각탑#1  
4349  2022-05-31  2022-05-31  2019-09-17 오후 3:40:40  냉각탑#1  
4410  2022-06-07  2022-06-07  2019-09-24 오후 4:55:30  냉각탑#1  
4416  2022-06-07  2022-06-07  2019-09-20 오후 1:45:54  냉각탑#1  
4418  2022-06-07  2022-06-07  2019-09-20 오후 1:45:54  냉각탑#1  
4602  2022-06-20  2022-06-20  2019-09-17 오후 3:40:40  냉각탑#1  
4758  2022-06-27  2022-06-27  2019-09-24 오후 4:55:30  냉각탑#1  
4877

## 작업시간 분석

In [43]:
fmsa10['job_jisi_dt'] = pd.to_datetime(fmsa10['job_jisi_dt'])
fmsa10['job_enddt'] = pd.to_datetime(fmsa10['job_enddt'])
fmsa10['time'] = fmsa10['job_enddt'] - fmsa10['job_jisi_dt']
fmsa10['time'].value_counts()

0 days     18395
1 days       415
2 days        72
3 days        71
4 days        39
5 days        16
8 days         7
6 days         6
7 days         4
10 days        4
11 days        4
13 days        4
-1 days        2
12 days        2
14 days        2
23 days        2
17 days        2
16 days        1
18 days        1
28 days        1
21 days        1
9 days         1
19 days        1
Name: time, dtype: int64

## 작업장비이력 / 장비마스터의 전체 장비 수 비교
* 40811개의 장비 중 3762개의 장비만 작업 이력이 있음

In [44]:
fmsx05 = pd.read_excel(path+'FMSX05_장비마스터.xlsx')

### 장비코드는 다르나 장비 이름이 같은 경우 존재

In [45]:
fmsx05['fac_cd'].value_counts()

200201040293    1
200105080004    1
200104010466    1
200104010746    1
200104010615    1
               ..
200201010700    1
200201010702    1
200201010703    1
200302120001    1
700201020001    1
Name: fac_cd, Length: 40811, dtype: int64

In [46]:
fmsx05['fac_nm'].value_counts()

L2스위치                             1621
test                               927
6동 분말소화기                           577
1동 분말소화기                           487
5동 분말소화기                           429
                                  ... 
(6)정풍량 #3F (A09-CAV2,3 2대)           1
(6)정풍량 #3F (A08-CAV2,3 5대)           1
(6)대-3-3F(남)365호옆-화장실) 대변기 2대        1
(6)대-3-3F(여)309호옆-화장실) 대변기 4대        1
SCC 7호기 권상기                          1
Name: fac_nm, Length: 22256, dtype: int64

In [47]:
print('작업된 장비 개수',len(fmsa11['fac_cd'].value_counts()))
print('전체 장비 개수',len(fmsx05['fac_cd'].value_counts()))

작업된 장비 개수 3762
전체 장비 개수 40811


## 자재 출고 이력 / 장비마스터의 전체 장비 수 비교
* 2263개의 장비 중 1727개의 자재만 출고 이력이 있음

In [48]:
fmsb03 =  pd.read_excel(path+'FMSB03_자재마스터.xlsx')
fmsb03.head()

comp_cd     mat_cd  class_cd  mat_nm  mat_type  mat_std_qty  mat_cost  \
0    SEJO  101001001    101001    황동레일       NaN          NaN       NaN   
1    SEJO  101001002    101001  파이프서포트       NaN          NaN       NaN   
2    SEJO  101001003    101001   단관파이프       NaN          NaN       NaN   
3    SEJO  101001004    101001    각파이프       NaN          NaN       NaN   
4    SEJO  101003001    101003    초경철판       NaN          NaN       NaN   

  mat_unit  mat_maker  mat_country  ...  input_emp_no chg_emp_no input_scr_nm  \
0        개        NaN          NaN  ...           PJS        NaN          SQL   
1        개        NaN          NaN  ...           PJS        NaN          SQL   
2        본        NaN          NaN  ...           PJS        NaN          SQL   
3        개        NaN          NaN  ...           PJS        NaN          SQL   
4        매        NaN          NaN  ...           PJS        NaN          SQL   

   use_yn del_yn  old_mat_cd new_mat_cd old_class_cd new_class_cd  mat_yn  
0       Y      N   718006001  101001001      0718006       101001     NaN  
1       Y      N   101021002  101001002      0101021       101001     NaN  
2       Y      N   605002011  101001003      0605002       101001     NaN  
3       Y      N   101002014  101001004      0101002       101001     NaN  
4       Y      N   117011002  101003001      0117011       101003     NaN  

[5 rows x 24 columns]

### 자재코드는 다르나 자재 이름이 같은 경우 존재

In [49]:
fmsb03['mat_cd'].value_counts()

101001001     1
415004001     1
415003003     1
415003004     1
415003005     1
             ..
207001028     1
207001027     1
207001026     1
207001025     1
1107002002    1
Name: mat_cd, Length: 2263, dtype: int64

In [50]:
fmsb03['mat_nm'].value_counts()

커넥터         6
주차관제주변기기    5
냉각팬         5
파워서플라이      5
스피커         3
           ..
도로표지용도료     1
니스붓         1
회전문 부품      1
고임목         1
눈삽          1
Name: mat_nm, Length: 2171, dtype: int64

In [51]:
print('출고된 자재 개수',len(fmsb07['mat_cd'].value_counts()))
print('전체 자재 개수',len(fmsb03['mat_cd'].value_counts()))

출고된 자재 개수 1727
전체 자재 개수 2263


## 수리이력이 많은 장비 분석
* 특정장비가 정비가 많은 경우 파악이 가능하나 그 외의 뚜렷한 특징은 파악하지 못함

In [52]:
fmsa11['fac_cd'].value_counts()

200101020028    24
200101020029    24
200101020002    24
200101020003    23
200101020005    22
                ..
300401020004     1
400301020257     1
300201011251     1
400301020185     1
200201030273     1
Name: fac_cd, Length: 3762, dtype: int64

In [53]:
# 냉각탑1 13동 6층 옥상
equip_work2[equip_work2['fac_cd'] == 200101020028]

index comp_cd        job_cd  job_tpy  \
130     254    SEJO  211006-00043        3   
361     716    SEJO  211014-00039      122   
3799  11034    SEJO  220426-00036        3   
3963  11506    SEJO  220504-00037        3   
4369  13040    SEJO  220602-00029        3   
4413  13231    SEJO  220607-00050        3   
4507  13572    SEJO  220613-00002        3   
4605  13966    SEJO  220620-00020        3   
4760  14442    SEJO  220627-00040        3   
4887  14865    SEJO  220704-00039        3   
5043  15322    SEJO  220711-00025        3   
5165  15752    SEJO  220718-00017        3   
5288  16187    SEJO  220725-00031        3   
5389  16577    SEJO  220801-00038        3   
5542  17024    SEJO  220808-00013        3   
5704  17520    SEJO  220816-00017        3   
5812  17865    SEJO  220822-00035        3   
6079  18687    SEJO  220902-00025        3   
6096  18777    SEJO  220905-00005        3   
6098  18779    SEJO  220905-00063        3   
6244  19164    SEJO  220913-00023        3   
6393  19525    SEJO  220919-00031        3   
6559  19967    SEJO  220926-00027        3   
6709  20395    SEJO  221004-00081        3   

                                    job_nm  job_stat job_jisi_dt   job_enddt  \
130        13-1동 6층 냉각탑 1,2호기 냉각수처리제 6통 투입         4  2021-10-06  2021-10-06   
361                13-1동 냉각탑 퇴수 작업 및 청소 작업         4  2021-10-14  2021-10-14   
3799             13동 6층 냉각탑 수조청소 준비 및 상태점검         4  2022-04-26  2022-04-26   
3963     13-1동 냉각탑 1,2호기 내부수조 청소 및 수처리제 투입         4  2022-05-04  2022-05-04   
4369  13동 옥상 냉각탑 1,2호기 수처리제 워터가드 1KG 6통 투입         4  2022-06-02  2022-06-02   
4413  13동 옥상 냉각탑 1,2호기 수처리제 워터가드 1KG 6통 투입         4  2022-06-07  2022-06-07   
4507          13-1동 6층 냉각탑 1,2호기 냉각수처리제 투입         4  2022-06-13  2022-06-13   
4605  13-1동 냉각탑 1,2호기 내부수조 청소 및 수처리제 3통 투입         4  2022-06-20  2022-06-20   
4760    13동 옥상 냉각탑 1호기 수처리제 워터가드 1KG 3통 투입         4  2022-06-27  2022-06-27   
4887  13동 옥상 냉각탑 1,2호기 수처리제 워터가드 1KG 6통 투입         4  2022-07-04  2022-07-04   
5043          13-1동 6층 냉각탑 1,2호기 냉각수처리제 투입         4  2022-07-11  2022-07-11   
5165  13-1동 냉각탑 1,2호기 내부수조 청소 및 수처리제 6통 투입         4  2022-07-18  2022-07-18   
5288  13동 옥상 냉각탑 1,2호기 수처리제 워터가드 1KG 6통 투입         4  2022-07-25  2022-07-25   
5389  13동 옥상 냉각탑 1,2호기 수처리제 워터가드 1KG 6통 투입         4  2022-08-01  2022-08-01   
5542          13-1동 6층 냉각탑 1,2호기 냉각수처리제 투입         4  2022-08-08  2022-08-08   
5704             13-1동 냉각탑 1,2호기 냉각수처리제 투입         4  2022-08-16  2022-08-16   
5812  13동 옥상 냉각탑 1,2호기 수처리제 워터가드 1KG 6통 투입         4  2022-08-22  2022-08-22   
6079         13-1동 6층 옥상 냉각탑 1,2호기 수처리제 투입         4  2022-09-02  2022-09-02   
6096          13-1동 6층 냉각탑 1,2호기 냉각수처리제 투입         4  2022-09-05  2022-09-05   
6098               13-1동 6층 냉각탑 1,2호기 수질검사         4  2022-09-05  2022-09-05   
6244                13-1동 6층 냉각탑 냉각수처리제 투입         4  2022-09-13  2022-09-13   
6393         13-1동 6층 옥상 냉각탑 1,2호기 수처리제 투입         4  2022-09-19  2022-09-19   
6559          13-1동 냉각탑 1,2호기 냉각수처리제 6통 투입         4  2022-09-26  2022-09-26   
6709          13-1동 냉각탑 1,2호기 냉각수처리제 6통 투입         4  2022-10-04  2022-10-04   

                     job_shm                 job_ehm  ... common_wrk emc_id  \
130    2021-10-06 오전 9:30:00  2021-10-06 오전 10:30:00  ...        NaN    NaN   
361   2021-10-14 오전 10:00:00   2021-10-14 오후 4:00:00  ...        NaN    NaN   
3799   2022-04-26 오후 1:00:00   2022-04-26 오후 3:00:00  ...        NaN    NaN   
3963   2022-05-04 오후 1:00:00   2022-05-04 오후 3:00:00  ...        NaN    NaN   
4369   2022-06-02 오전 9:30:00  2022-06-02 오전 10:30:00  ...        NaN    NaN   
4413   2022-06-07 오전 9:30:00  2022-06-07 오전 10:30:00  ...        NaN    NaN   
4507   2022-06-13 오전 9:00:00  2022-06-13 오전 10:00:00  ...        NaN    NaN   
4605   2022-06-20 오전 9:30:00  2022-06-20 오전 10:30:00  ...        NaN    NaN   
4760   2022-06-27 오전 9:30:00  2022-06-27 오전 10:30:00  ...        NaN    NaN   
4887   2022-07-04 오전 9:30:00  2022-07-04 오전 10:30:00  ...       

## 관련이 있을거 같은 컬럼 찾기
* 'A'컬럼에 데이터가 입력이 되어 있으면 'B'컬럼에도 데이터 입력이 되어있을것 > Null값 개수 같을 것
  >>(job_ilbo,job_fac_his) / (chg_sysdt,chg_emp_no)

In [54]:
col_list = fmsa10.columns
for i in col_list:
    print(i, ' : ', fmsa10[i].isnull().sum())

comp_cd  :  0
job_cd  :  0
job_tpy  :  0
job_nm  :  0
job_stat  :  0
job_jisi_dt  :  0
job_enddt  :  7
job_shm  :  1
job_ehm  :  7
job_cause  :  19036
job_cont  :  1503
job_ilbo  :  1288
job_kpi  :  7
job_fac_his  :  1288
job_mng  :  19060
day_night_ind  :  7
dept_cd  :  0
class_cd  :  7
sjob_cd  :  17294
vend_nm  :  18876
vend_amt  :  19038
as_req_man  :  12557
as_req_tel  :  12006
as_req_dt  :  2329
as_jubsu  :  2331
input_sysdt  :  0
chg_sysdt  :  4821
input_emp_no  :  0
chg_emp_no  :  4821
input_scr_nm  :  0
msg_yn  :  0
as_req_cont  :  2629
as_req_comp  :  11864
as_happycall_dt  :  19060
as_happycall_man  :  19060
as_happycall_res1  :  18516
as_happycall_res2  :  18516
as_happycall_score  :  19060
as_jungyo  :  0
as_submit  :  14581
as_reg_h  :  11656
as_reg_d  :  10526
gyuljae_no1  :  9130
gyuljae_no2  :  19060
gyuljae_no3  :  19060
gyuljae_no4  :  19060
gyuljae_no5  :  19060
gyuljae_yn  :  0
edit_yn  :  0
job_doc_yn  :  0
sjob_jisi_dt  :  17294
work_time  :  19060
bosu_gyuljae_y